rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/ContextMenu.java

```diff
@@ -0,0 +1,184 @@
+        lockItem.addClickHandler(event14 -> controller.lockNodes());
+
+        MenuItem unlockItem = new MenuItem("Unlock", menu.unlockItemImageResource);
+        unlockItem.addClickHandler(event13 -> controller.unlockNodes());
+
+        MenuItem deployItem = new MenuItem("Deploy", menu.deployItemImageResource);
+        deployItem.addClickHandler(event12 -> controller.deployNodeSource());
+
+        MenuItem undeployItem = new MenuItem("Undeploy", menu.undeployItemImageResource);
+        undeployItem.addClickHandler(event1 -> controller.undeployNodeSource());
+
+        MenuItem editItem = new MenuItem("Edit", menu.editItemImageResource);
+
+        String nodeSourceName = menu.nodesource == null ? "" : menu.nodesource.getSourceName();
+        NodeSourceStatus nodeSourceStatus = menu.nodesource == null ? null : menu.nodesource.getNodeSourceStatus();
+        editItem.addClickHandler(event1 -> controller.editNodeSource(nodeSourceName, nodeSourceStatus));
+
+        if (menu.node != null) {
+            if (menu.node.isLocked()) {
+                lockItem.setEnabled(false);
+                unlockItem.setEnabled(true);
+            } else {
+                lockItem.setEnabled(true);
+                unlockItem.setEnabled(false);
+            }
+        }
+
+        if (menu.nodesource != null) {
+            switch (menu.nodesource.getNodeSourceStatus()) {
+                case NODES_DEPLOYED:
+                    editItem.setTitle(EditDynamicParametersWindow.WINDOW_TITLE);
+                    menu.enableItems(undeployItem);
+                    menu.disableItems(deployItem);
+                    break;
+                case NODES_UNDEPLOYED:
+                    editItem.setTitle(EditNodeSourceWindow.WINDOW_TITLE);
+                    menu.enableItems(deployItem);
+                    menu.disableItems(undeployItem);
+                    break;
+                default:
+                    menu.disableItems(deployItem, undeployItem, editItem);
+            }
+        } else {
+            menu.disableItems(deployItem, undeployItem, editItem);
+        }
+
+        menu.setItems(deployItem, undeployItem, editItem, new MenuItemSeparator(), lockItem, unlockItem, removeItem);
+
+        return menu;
+    }
+
+    private void init(Object related) {
+        if (related instanceof NodeSource) {
+            nodesource = (NodeSource) related;
+        } else if (related instanceof NodeSource.Host) {
+            host = (NodeSource.Host) related;
+        } else if (related instanceof NodeSource.Host.Node) {
+            node = (NodeSource.Host.Node) related;
+        } else {
+            LogModel.getInstance().logCriticalMessage("Cannot show context menu, related object is wrong type: " +
+                                                      related.getClass().getCanonicalName());
+            throw new RuntimeException("Cannot show context menu, related object is wrong type: " +
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)